backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/migration/MigrationTask.java

```diff
@@ -43,29 +76,73 @@ public int getToVersion() {
         }
     }
 
-    private final ImmutableList<Migration> migrations;
-    private final SchemaVersion toVersion;
+    private final CassandraSchemaVersionDAO schemaVersionDAO;
+    private final CassandraSchemaTransitions transitions;
+    private final SchemaVersion target;
 
-    public MigrationTask(List<Migration> migration, SchemaVersion toVersion) {
-        this.migrations = ImmutableList.copyOf(migration);
-        this.toVersion = toVersion;
+    @VisibleForTesting
+    public MigrationTask(CassandraSchemaVersionDAO schemaVersionDAO, CassandraSchemaTransitions transitions, SchemaVersion target) {
+        this.schemaVersionDAO = schemaVersionDAO;
+        this.transitions = transitions;
+        this.target = target;
     }
 
     @Override
-    public Result run() throws InterruptedException {
-        for (Migration migration: migrations) {
-            migration.asTask().run();
-        }
+    public Result run() {
+        getCurrentVersion().listTransitionsForTarget(target)
+                .stream()
+                .map(this::migration)
+                .forEach(Throwing.consumer(this::runMigration).sneakyThrow());
         return Result.COMPLETED;
     }
 
+    private SchemaVersion getCurrentVersion() {
+        return schemaVersionDAO.getCurrentSchemaVersion().block().orElse(DEFAULT_VERSION);
+    }
+
+    private Tuple2<SchemaTransition, Migration> migration(SchemaTransition transition) {
+        return Tuples.of(
+                transition,
+                transitions.findMigration(transition)
+                        .orElseThrow(() -> new MigrationException("unable to find a required Migration for transition " + transition)));
+
+    }
+
+    private void runMigration(Tuple2<SchemaTransition, Migration> tuple) throws InterruptedException {
+        SchemaVersion currentVersion = getCurrentVersion();
+        SchemaTransition transition = tuple.getT1();
+        SchemaVersion targetVersion = transition.to();
+        if (currentVersion.isAfterOrEquals(targetVersion)) {
+            return;
+        }
+
+        LOGGER.info("Migrating to version {} ", transition.toAsString());
+        Migration migration = tuple.getT2();
+        migration.asTask().run()
+                .onComplete(
+                        () -> schemaVersionDAO.updateVersion(transition.to()).block(),
```

Comment:
I didn't changed that, it's an choice made in the past.
However, I think it was done on purpose: the Migration is responsible for changing the data and this infrastructure code is responsible for version management.
As tasks are idempotent, if a task fails it won't update the version. Next time we'll run the same task, it will have the same result as if it wasn't interrupted. Only then, it'll upgrade the  version. It seems fin to me.
Do you have some corner case in mind?

Reply:
I can think of one. 
The task is requested to run on a node A. The on complete listener is attached on this node.
The task run on a node B.
The node A fail/restart.
The task complete successfully on B.
The on complete listener no longer exists on A. And so the version is never updated

Reply:
Or i am missing something ?

Reply:
The unit of execution here is MigrationTask. The subtask (that are tasks only for limiting the refactoring) are not run as tasks on the task manager so this won't happen.
I'm wondering if reusing Task concept without a Task Manager is not too misleading.
Do you think it's worth refactoring ? (anyway, It would be in another PR)

Reply:
Ok, yes i think opening a PR to remove this ambiguity would be a good thing.